EIPS/eip-7512.md

```diff
@@ -18,71 +18,89 @@ The proposal aims to create a standard for an onchain representation of audit re
-    - `ercs` - A list of ERCs that are implemented by the target contract. The ERCs listed MUST be fully implemented. This list MAY be empty.
-    - `auditHash` - MUST be the hash of the original audit. This allows to verify that this onchain information belongs to a specific audit.
-    - `auditUri` - SHOULD point to a source where the audit can be retrieved
+    - `auditor`: Information on the auditor
+    - `auditedContract`: MUST be the `chainId` as well as `implementation` of the deployed contract the audit is related to
+    - `issuedAt`: MUST contain the information when the original audit (identified by the `auditHash`) was issued
+    - `ercs`: A list of ERCs that are implemented by the target contract. The ERCs listed MUST be fully implemented. This list MAY be empty
+    - `auditHash`: MUST be the hash of the original audit. This allows onchain verification of information that may belong to a specific audit
+    - `auditUri`: SHOULD point to a source where the audit can be retrieved
+- Contract
+    - `chainId`: MUST be a `bytes32` representation of the [ERC-155](./eip-155.md) chain ID of the blockchain that the contract has been deployed in
+    - `implementation`: MUST be a `bytes32` representation of a contract's implementation as described below
+
+### Contract Implementation
+
+Security audits generally cover a specific snapshot of an overall codebase, meaning that consequent updates of the code are not to be considered covered by it. While smart contracts are immutable on a blockchain, multiple EIP standards have been defined (i.e. ERC, ERC, etc.) that permit a "circumvention" of this blockchain trait and allow the logic of a smart contract to be updated. 
+
+In this regard, simply associating a security audit with an `address` of a smart contract is insufficient as it may ultimately change its underlying implementation. In order to future-proof this EIP and cover certain cases such as polymorphic contracts that permit a contract to be re-deployed in the same `address`, we provide the following definition of a contract `implementation`.
+
+An `implementation` is a `bytes32` representation of a `keccak256` hash-chain of the contract's `address` and bytecode hashes derived using the `EXTCODEHASH` opcode ([ERC-1052](./eip-1052.md)). To elaborate, each "contract" can be thought of as consisting of an `address` that it is deployed in as well as its bytecode. When dealing with proxies, diamond patterns, and other such approaches we end up with "contracts" that ultimately relay calls to other "contracts".
+
+As a function's execution will ultimately end up at a final "contract" that represents the logic executed, we can think of the `implementation` hash-chain as being composed of atomic "contract" hashes that are chained together until the destination logic address. Below is a pseudo-code snippet of how this would look in practice:
+
+```solidity
+address proxy;
+address implementation;
```

Comment:
The challenge here is that to properly compare this it would be necessary to load the `implementation` from the proxy. Not sure if this is even always possible (i.e. the [minimal proxy implementation](https://eips.ethereum.org/EIPS/eip-1167) does not allow this).

Also diamond proxies allow "partial" reuse of the implementation contract (per function based), which has a security impact too. 

Therefore my motivation was to handle proxies outside of this ERC in a first version. I.e. there could be registries that are proxy aware. So they use the audits and then check for specific proxy implementations if they use that contract mentioned in the audit as a implementation. 

```solidity
function build1167ProxyCodehash(address proxy) internal returns (bytes32);

function checkAuditForProxy(address proxy, AuditSummary audit) public {
  required(proxy.codehash == build1167ProxyCodehash(audit.address), "Unsupported proxy");
  
  // verify AuditSummary
}
```

Reply:
Thanks for the feedback @rmeissner! I have included a disclaimer in relation to Diamond standards in the MD later on. In regard to the minimal proxy, **it does actually allow it** albeit using byte indices and loading the code of the deployed address in memory. Given that it is only a few bytes, it is trivial to implement a function that extracts the address.

As we also specify ERC support when submitting an audit, we can simply submit the proxy along with the proxy standard it supports and allow the ecosystem to build proper methods for extracting the proxy implementation address to validate audits.

We do not need to define the actual way the ecosystem extracts the necessary data entries to construct the `bytes32 implementation`, however, I think we should specify a future-proof way to handle how an audit is associated with a contract instance which can take many forms.

Reply:
Wouldn't it make more sense then to add more explicit proxy information to the audit summary?

I.e. a boolean if it is a proxy, if it is upgradable and a name of the proxy (to indicate how to load the implementation address onchain). 

My challenge here is that there are quite some proxy contracts (with and without standards) and I am not sure how to cover all of them, in a concise way.

Reply:
The standard itself will cover them regardless of whether they comply with known standards or not. At its core, it is a hash chain of `address`, `runtimeCode` consecutively. By definition, **all proxies** in any shape or form need to comply with this paradigm due to the usage of `delegatecall` (which is the only instruction that permits proxies in the first place).

How the ecosystem ends up actually extracting this information on-chain should not be a concern of the ERC. In regard to off-chain extraction of the implementation, it is much more trivial and could brute-force known standards or even fetch past transactions and inspect their call chain to see what addresses were `delegatecall`-ed to.

Reply:
Ok, I think I am aligned with you now. Last minor nitpick is the name `implementation` (got me confused as the address usage is something different than the bytes32). It is like a `checksum` so using that as a name might be easier to understand.

Reply:
Good to hear, I have no objections to changing the name of the variable. I agree `implementation` is not the best description, `checksum` is more accurate and I am aligned with it. I will also propose `digest` which I think I've seen used in ERCs more often. Either is definitely better than `implementation` though as it may indeed mislead people at first glance.